### PR TITLE
Add getset_descriptor

### PR DESCRIPTION
**Pull request by youknowone**
- Add getset_descriptor
- Remove PyReadOnlyProperty
- #[pyproperty] now generates getset. PyProperty is only used for python code.

### CODE REVIEW

**Comment by coolreader18** [comment]
How is PyGetSet different from PyProperty?

**Comment by youknowone** [comment]
in CPython, PyGetSet corresponds to `getset_descriptor` while PyProperty corresponds to `property`. this is data specialized form of `property`. How about thinking like `PyBulitinMethod` and `PyBoundMethod` relationship?

Practically, it only corresponds with native functions with solid signature. So it must be related to performance. For compatibility point of view, the properties of `None` object matters too. Because descriptors get obj argument for `__get__` and checking its existance by not-equal-None, properties of `None` object is nessessarily broken. `getset_descriptor` doesn't check it but always call getter function.

**Comment by coolreader18** [comment]
Maybe it would be good to replace all the property APIs (`new_property`, `PropertyBuilder`, etc) with these descriptors, if not in this PR then in another one. It seems like there's not really a benefit to using properties from native code, so in basically every situation you'd prefer getset over property (just not in, like, the `builtins.property()` decorator)

**Comment by youknowone** [comment]
After grep CPython code, yes, that sounds true.

**Comment by youknowone** [comment]
One problem left. I have no idea how CPython shows descriptor for following code while RustPython raises error.

```python
>>> Exception.args
<attribute 'args' of 'BaseException' objects>
```

File: vm/src/obj/objgetset.rs [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+    }
+}
+
+pub type PyGetSetRef = PyRef<PyGetSet>;
+
+impl PyBuiltinDescriptor for PyGetSet {
+    fn get(
+        zelf: PyRef<Self>,
+        obj: PyObjectRef,
+        _cls: OptionalArg<PyObjectRef>,
```
Comment (coolreader18) [inline vm/src/obj/objgetset.rs]: Maybe to fix the exception.args case, you could have this check `cls` and if it's not there than just return `zelf`?

Comment (youknowone) [reply vm/src/obj/objgetset.rs]: I copied `descr_check` part of CPython. Now `descr_get` signature looks a little bit messy than before

**Comment by youknowone** [comment]
@coolreader18 It seems this is finally done! will you review it please?

File: vm/src/stdlib/io.rs [hunk 2]
```diff
@@ -573,7 +573,7 @@ mod fileio {
         let fileno = i64::try_from_object(vm, vm.get_attribute(instance.clone(), "__fileno")?)?;
         Ok(os::rust_file(fileno))
     }
-    fn fio_set_fileno(instance: &PyObjectRef, f: fs::File, vm: &VirtualMachine) -> PyResult<()> {
+    fn fio_set_fileno(instance: &PyObjectRef, f: fs::File, vm: &VirtualMachine) -> PySetResult {
```
[the remark this answers is not in the document]
Comment (youknowone) [reply vm/src/stdlib/io.rs]: I changed the name `IntoPySetResult` to `IntoPyNoResult` and removed `PySetResult` type.

**Comment by youknowone** [comment]
I attached #1736 to here because it is trivial derived work from this PR

**Comment by youknowone** [comment]
Other getters of SyntaxError are actually not a getter by args. they are attributes set by `set_attr`

**Pull request merged**